php/libraries/NDB_Caller.class.inc

```diff
@@ -443,32 +443,42 @@ class NDB_Caller
         $commentID,
         $redirectToOnSuccess=null
     ) {
-        if ($page === 'finalpage') {
-            $instrument =& NDB_BVL_Instrument::factory(
-                $instrumentName,
-                $commentID,
-                $page
-            );
+        $instrument = NDB_BVL_Instrument::factory(
+            $instrumentName,
+            $commentID,
+            $page
+        );
 
-            $this->instrument = $instrument;
+        $this->instrument = $instrument;
+        if (is_null($instrument)) {
+            throw new Exception("Instrument does not exist");
+        }
+        if ($page === 'finalpage') {
             return "<script language='javascript'>"
                 . "window.location='$redirectToOnSuccess'"
                 . "</script>";
         }
         $user =& User::singleton();
 
-        if ($_REQUEST['sessionID']) {
-            $timepoint =& TimePoint::singleton($_REQUEST['sessionID']);
-        }
-
-        // make an instance of the instrument's object
-        $instrument =& NDB_BVL_Instrument::factory(
-            $instrumentName,
-            $commentID,
-            $page
-        );
+        $session_id = $instrument->getSessionID();
+        $timepoint  = TimePoint::singleton($session_id);
 
-        $this->instrument = $instrument;
+        // This prevents users from requesting an unassigned sessionID for
+        // a candidate and consequently writing to it.
+        if (is_null($timepoint)) {
+            throw new LorisException(
+                "Time point does not exist for session {$session_id}"
+            );
+        }
+        // This preserves backwards compatibility for any code
+        // that will be called that assumes
+        // $_REQUEST["sessionID"] will be populated
+        $_REQUEST["sessionID"] = $session_id;
+        if (strtoupper($_SERVER["REQUEST_METHOD"]) === "GET") {
+            $_GET["sessionID"] = $session_id;
+        } else {
```

Comment:
My preference is to change this to an `else if` and explicitly check for `POST` just to be robust. Not sure if other request methods are used in the codebase.